File: src/main/java/com/kingmang/lazurite/parser/AST/Expressions/TernaryExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.values.LzrValue;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public final class TernaryExpression implements Expression {
    public final Expression condition;
    public final Expression trueExpr, falseExpr;

    @Override
    public LzrValue eval() {
        if (condition.eval().asInt() != 0) {
            return trueExpr.eval();
        } else {
            return falseExpr.eval();
        }
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return String.format("%s ? %s : %s", condition, trueExpr, falseExpr);
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/TernaryExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.values.LzrValue

data class TernaryExpression(
    val condition: Expression,
    val trueExpr: Expression,
    val falseExpr: Expression
) : Expression {
    override fun eval(): LzrValue =
        if (condition.eval().asInt() != 0)
            trueExpr.eval()
        else
            falseExpr.eval()

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        String.format("%s ? %s : %s", condition, trueExpr, falseExpr)
}
